PCA - Principal Component Analysis  
jest techniką zmniejszającą ilość wymiarów.   
Polega na określeniu najważniejszych cech, oraz na określeniu ważności wsyzstkich cech,  
które mają największy wpływ na zmienną docelową   

Cechy muszą być przeskalowane

Skutkiem PCA jest szybszy czas uczenia, kosztem precyzji

In [2]:
import pandas as pd
from sklearn.datasets import load_digits

In [3]:
digits = load_digits()

In [4]:
df = pd.DataFrame(digits.data,columns=digits.feature_names)
df.describe()

pixel_0_0    pixel_0_1    pixel_0_2    pixel_0_3    pixel_0_4  \
count     1797.0  1797.000000  1797.000000  1797.000000  1797.000000   
mean         0.0     0.303840     5.204786    11.835838    11.848080   
std          0.0     0.907192     4.754826     4.248842     4.287388   
min          0.0     0.000000     0.000000     0.000000     0.000000   
25%          0.0     0.000000     1.000000    10.000000    10.000000   
50%          0.0     0.000000     4.000000    13.000000    13.000000   
75%          0.0     0.000000     9.000000    15.000000    15.000000   
max          0.0     8.000000    16.000000    16.000000    16.000000   

         pixel_0_5    pixel_0_6    pixel_0_7    pixel_1_0    pixel_1_1  ...  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  ...   
mean      5.781859     1.362270     0.129661     0.005565     1.993879  ...   
std       5.666418     3.325775     1.037383     0.094222     3.196160  ...   
min       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
25%       0.000000     0.000000     0.000000     0.000000     0.000000  ...   
50%       4.000000     0.000000     0.000000     0.000000     0.000000  ...   
75%      11.000000     0.000000     0.000000     0.000000     3.000000  ...   
max      16.000000    16.000000    15.000000     2.000000    16.000000  ...   

         pixel_6_6    pixel_6_7    pixel_7_0    pixel_7_1    pixel_7_2  \
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000   
mean      3.725097     0.206455     0.000556     0.279354     5.557596   
std       4.919406     0.984401     0.023590     0.934302     5.103019   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     1.000000   
50%       1.000000     0.000000     0.000000     0.000000     4.000000   
75%       7.000000     0.000000     0.000000     0.000000    10.000000   
max      16.000000    13.000000     1.000000     9.000000    16.000000   

         pixel_7_3    pixel_7_4    pixel_7_5    pixel_7_6    pixel_7_7  
count  1797.000000  1797.000000  1797.000000  1797.000000  1797.000000  
mean     12.089037    11.809126     6.764051     2.067891     0.364496  
std       4.374694     4.933947     5.900623     4.090548     1.860122  
min       0.000000     0.000000     0.000000     0.000000     0.000000  
25%      11.000000    10.000000     0.000000     0.000000     0.000000  
50%      13.000000    14.000000     6.000000     0.000000     0.000000  
75%      16.000000    16.000000    12.000000     2.000000     0.000000  
max      16.000000    16.000000    16.000000    16.000000    16.000000  

[8 rows x 64 columns]

In [5]:
y = digits.target

In [6]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_scaled = scaler.fit_transform(df)

In [7]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X_scaled,y,test_size=0.2,random_state=30)

In [8]:
from sklearn.linear_model import LogisticRegression
model = LogisticRegression()
model.fit(X_train,y_train)
model.score(X_test,y_test)

0.9722222222222222

In [9]:
from sklearn.decomposition import PCA
pca = PCA(0.95) # Bierze 95% najbardziej przydatnych cech i tworzy nowe wymiary, ( 95% informaacji )
X_pca = pca.fit_transform(df)
X_pca.shape

(1797, 29)

**W tym przypadku nie wybiera 29 kolumn, tylko je wylicza**

pca.explained_variance_ratio_ zwraca listę, której index odpowiada jednej kolumnie.  
przykładowo 0.14890594 oznacza, że ta kolumna zawiera tyle procent informacji z df

In [10]:
pca.explained_variance_ratio_

array([0.14890594, 0.13618771, 0.11794594, 0.08409979, 0.05782415,
       0.0491691 , 0.04315987, 0.03661373, 0.03353248, 0.03078806,
       0.02372341, 0.02272697, 0.01821863, 0.01773855, 0.01467101,
       0.01409716, 0.01318589, 0.01248138, 0.01017718, 0.00905617,
       0.00889538, 0.00797123, 0.00767493, 0.00722904, 0.00695889,
       0.00596081, 0.00575615, 0.00515158, 0.0048954 ])

In [11]:
pca.n_components_

29

In [13]:
X_train,X_test,y_train,y_test = train_test_split(X_pca,y,test_size=0.2,random_state=30)

In [14]:
model_pca = LogisticRegression(max_iter=200)
model_pca.fit(X_train,y_train) 
model_pca.score(X_test,y_test)

0.9722222222222222

Można też przy tworzeniu pca określić liczbę wymiarów

In [15]:
pca = PCA(n_components=2)
X_pca = pca.fit_transform(df)
X_pca.shape

(1797, 2)

In [16]:
X_pca

array([[ -1.2594663 ,  21.27488434],
       [  7.95761099, -20.76869753],
       [  6.99192255,  -9.95598775],
       ...,
       [ 10.80128668,  -6.96025411],
       [ -4.87210435,  12.42395649],
       [ -0.34438732,   6.36554641]])

In [17]:
pca.explained_variance_ratio_

array([0.14890594, 0.13618771])

Jak można zauyważyć, zmniejszając model do dwóch cech, mamy w obydwu kolumnach w sumie mamy około 28% informacji  
Po wytrenowaniu modelu jego dokładnośc w tym przypadku wynosiłaby około 60%